library/std/src/sys/windows/stdio.rs

```diff
@@ -29,6 +31,32 @@ struct IncompleteUtf8 {
     len: u8,
 }
 
+impl IncompleteUtf8 {
+    // Implemented for use in Stdin::read.
+    fn read(&mut self, buf: &mut [u8]) -> usize {
+        // Write to buffer until the buffer is full or we run out of bytes.
+        let to_write = cmp::min(buf.len(), self.len as usize);
+        for (ind, val) in self.bytes[..to_write].iter().enumerate() {
+            buf[ind] = *val;
+        }
+
+        // Rotate the remaining bytes if not enough remaining space in buffer.
+        if usize::from(self.len) > buf.len() {
+            let new_len = self.len - (buf.len() as u8);
+            let mut new_bytes: [u8; 4] = [0; 4];
+            for (ind, val) in self.bytes[to_write..].iter().enumerate() {
+                new_bytes[ind] = *val;
+            }
+            self.bytes = new_bytes;
+            self.len = new_len;
+        } else {
+            self.len = 0;
+        }
```

Comment:
This section can probably be simplified using [`[T]::copy_within`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.copy_within)

Reply:
E.g.:
```rust
self.buf.copy_within(to_write.., 0)
```